Sources/BraintreePayPal/BTPayPalReturnURL.swift

```diff
@@ -0,0 +1,86 @@
+import Foundation
+
+#if canImport(BraintreeCore)
+import BraintreeCore
+#endif
+
+enum BTPayPalAppSwitchReturnURLState {
+    case unknownPath
+    case succeeded
+    case canceled
+}
+
+/// This class interprets URLs received from the PayPal app via app switch returns.
+///
+/// PayPal app switch authorization requests should result in success or user-initiated cancelation. These states are communicated in the url.
+struct BTPayPalReturnURL {
+
+    /// The overall status of the app switch - success, cancelation, or an unknown path
+    var state: BTPayPalAppSwitchReturnURLState = .unknownPath
+
+    /// Initializes a new `BTPayPalAppSwitchReturnURL`
+    /// - Parameter url: an incoming app switch url
+    init?(_ redirectType: PayPalRedirectType) {
+        switch redirectType {
+        case .payPalApp(let url), .webBrowser(let url):
+            if url.path.contains("success") {
+                state = .succeeded
+            } else if url.path.contains("cancel") {
+                state = .canceled
+            } else {
+                state = .unknownPath
+            }
+        }
+    }
+
+    // MARK: - Static Methods
+
+    /// Evaluates whether the url represents a valid PayPal return URL.
+    /// - Parameter url: an app switch return URL
+    /// - Returns: `true` if the url represents a valid PayPal app switch return
+    static func isValid(_ url: URL) -> Bool {
+        url.scheme == "https" && (url.path.contains("cancel") || url.path.contains("success"))
+    }
+
+    static func isValidURLAction(url: URL, linkType: String?) -> Bool {
+        guard let host = url.host, let scheme = url.scheme, !scheme.isEmpty else {
+            return false
+        }
+
+        var hostAndPath = host
+            .appending(url.path)
+            .components(separatedBy: "/")
+            .dropLast(1) // remove the action (`success`, `cancel`, etc)
+            .joined(separator: "/")
+
+        if hostAndPath.count > 0 {
+            hostAndPath.append("/")
+        }
+
+        /// If we are using the deeplink/ASWeb based PayPal flow we want to check that the host and path matches
+        /// the static callbackURLHostAndPath. For the universal link flow we do not care about this check.
+        if hostAndPath != BTPayPalRequest.callbackURLHostAndPath && linkType == "deeplink" {
```

Comment:
@agedd I know you mentioned this logic for this specific if block being confusing before. I added the docstrings as well as passing/checking the `linkType` here as deeplink instead of passing in the request and doing some weird stuff with that. Please let me know if this makes it clearer what this check is doing!

Reply:
this looks great! @jaxdesmarais imo definitely a lot more clearer to read :)